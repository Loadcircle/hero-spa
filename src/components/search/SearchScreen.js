import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {

    const location = useLocation();

    const {q:query = ''} = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({heroName: query});
    const {heroName} = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(query), [query]);

    const handleSubmit = (e)=>{
        e.preventDefault();
        history.push(`?q=${heroName}`)
    }

    

    return (
        <div>
            <h1>
                Search Screen
            </h1>
            <hr/>
            <div className="row">

                <div className="col-5">
                    <h4> Search Form </h4>
                    <hr/>

                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            name="heroName"
                            value={heroName}
                            onChange={handleInputChange}
                        />

                        <button
                            type="submit"
                            className="btb mt-1 btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">

                    {
                        (heroesFiltered.length < 1)
                        ?
                        (
                            <div className="alert alert-warning">
                                Hero doesn't exist
                            </div>
                        )
                        :
                        (
                            <>
                                <h4> Results </h4>
                                <hr/>
                                {
                                    heroesFiltered.map(hero=>(
                                        <HeroCard 
                                            key={hero.id}
                                            {...hero}
                                        />
                                    ))
                                }
                            </>
                        )
                    }

                </div>

            </div>
        </div>
    )
}
