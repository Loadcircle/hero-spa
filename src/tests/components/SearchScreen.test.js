import { mount } from "enzyme";
import React from 'react'
import { MemoryRouter, Route } from "react-router-dom";
import { AuthContext } from "../../auth/AuthContext";
import { SearchScreen } from "../../components/search/SearchScreen";

describe('Pruebas en SearchScreen', ()=>{

    const historyMock = {
        push: jest.fn()
    }

    test('Debe de mostrarse correctamente con valores por defecto', ()=>{

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen}>

                </Route>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();

    });

    test('Debe de mostrar a Batman y el input con el valor del query', ()=>{

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen}>

                </Route>
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');

    });

    test('Debe de mostrar un erro si no hay heroes', ()=>{
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=sdfsdfsd']}>
                <Route path="/search" component={SearchScreen}>

                </Route>
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-warning').exists()).toBe(true);

    });

    test('Debe llamar al push del history', ()=>{
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    path="/search" 
                    component={()=> <SearchScreen history={historyMock}/>}>
                </Route>
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: 'searchText',
            value: 'batman'
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`);

    });

});