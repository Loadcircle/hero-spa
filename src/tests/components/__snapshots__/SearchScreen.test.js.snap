// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="heroName"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btb mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                   
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Superman"
              className="card-img"
              src="./assets/heroes/dc-superman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Superman
              </h5>
              <p
                className="card-text"
              >
                Kal-El
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Action Comics #1
                   
                </small>
              </p>
              <a
                href="/hero/dc-superman"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Flash"
              className="card-img"
              src="./assets/heroes/dc-flash.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Flash
              </h5>
              <p
                className="card-text"
              >
                Jay Garrick, Barry Allen, Wally West, Bart Allen
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Flash Comics #1
                   
                </small>
              </p>
              <a
                href="/hero/dc-flash"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Green Lantern"
              className="card-img"
              src="./assets/heroes/dc-green.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Green Lantern
              </h5>
              <p
                className="card-text"
              >
                Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  All-American Comics #16
                   
                </small>
              </p>
              <a
                href="/hero/dc-green"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Green Arrow"
              className="card-img"
              src="./assets/heroes/dc-arrow.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Green Arrow
              </h5>
              <p
                className="card-text"
              >
                Oliver Queen
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  More Fun Comics #73
                   
                </small>
              </p>
              <a
                href="/hero/dc-arrow"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Wonder Woman"
              className="card-img"
              src="./assets/heroes/dc-wonder.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Wonder Woman
              </h5>
              <p
                className="card-text"
              >
                Princess Diana
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  All Star Comics #8
                   
                </small>
              </p>
              <a
                href="/hero/dc-wonder"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Martian Manhunter"
              className="card-img"
              src="./assets/heroes/dc-martian.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Martian Manhunter
              </h5>
              <p
                className="card-text"
              >
                Martian Manhunter
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #225
                   
                </small>
              </p>
              <a
                href="/hero/dc-martian"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Robin/Nightwing"
              className="card-img"
              src="./assets/heroes/dc-robin.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Robin/Nightwing
              </h5>
              <p
                className="card-text"
              >
                Dick Grayson
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #38
                   
                </small>
              </p>
              <a
                href="/hero/dc-robin"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Blue Beetle"
              className="card-img"
              src="./assets/heroes/dc-blue.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Blue Beetle
              </h5>
              <p
                className="card-text"
              >
                Dan Garret, Ted Kord, Jaime Reyes
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Mystery Men Comics #1
                   
                </small>
              </p>
              <a
                href="/hero/dc-blue"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Black Canary"
              className="card-img"
              src="./assets/heroes/dc-black.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Black Canary
              </h5>
              <p
                className="card-text"
              >
                Dinah Drake, Dinah Lance
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Flash Comics #86
                   
                </small>
              </p>
              <a
                href="/hero/dc-black"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Spider Man"
              className="card-img"
              src="./assets/heroes/marvel-spider.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Spider Man
              </h5>
              <p
                className="card-text"
              >
                Peter Parker
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Amazing Fantasy #15
                   
                </small>
              </p>
              <a
                href="/hero/marvel-spider"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Captain America"
              className="card-img"
              src="./assets/heroes/marvel-captain.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Captain America
              </h5>
              <p
                className="card-text"
              >
                Steve Rogers
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Captain America Comics #1
                   
                </small>
              </p>
              <a
                href="/hero/marvel-captain"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Iron Man"
              className="card-img"
              src="./assets/heroes/marvel-iron.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Iron Man
              </h5>
              <p
                className="card-text"
              >
                Tony Stark
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Tales of Suspense #39
                   
                </small>
              </p>
              <a
                href="/hero/marvel-iron"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Thor"
              className="card-img"
              src="./assets/heroes/marvel-thor.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Thor
              </h5>
              <p
                className="card-text"
              >
                Thor Odinson
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Journey into Myster #83
                   
                </small>
              </p>
              <a
                href="/hero/marvel-thor"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Hulk"
              className="card-img"
              src="./assets/heroes/marvel-hulk.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Hulk
              </h5>
              <p
                className="card-text"
              >
                Bruce Banner
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  The Incredible Hulk #1
                   
                </small>
              </p>
              <a
                href="/hero/marvel-hulk"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Wolverine"
              className="card-img"
              src="./assets/heroes/marvel-wolverine.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Wolverine
              </h5>
              <p
                className="card-text"
              >
                James Howlett
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  The Incredible Hulk #180
                   
                </small>
              </p>
              <a
                href="/hero/marvel-wolverine"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Daredevil"
              className="card-img"
              src="./assets/heroes/marvel-daredevil.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Daredevil
              </h5>
              <p
                className="card-text"
              >
                Matthew Michael Murdock
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Daredevil #1
                   
                </small>
              </p>
              <a
                href="/hero/marvel-daredevil"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Hawkeye"
              className="card-img"
              src="./assets/heroes/marvel-hawkeye.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Hawkeye
              </h5>
              <p
                className="card-text"
              >
                Clinton Francis Barton
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Tales of Suspense #57
                   
                </small>
              </p>
              <a
                href="/hero/marvel-hawkeye"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Cyclops"
              className="card-img"
              src="./assets/heroes/marvel-cyclops.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Cyclops
              </h5>
              <p
                className="card-text"
              >
                Scott Summers
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  X-Men #1
                   
                </small>
              </p>
              <a
                href="/hero/marvel-cyclops"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Silver Surfer"
              className="card-img"
              src="./assets/heroes/marvel-silver.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="cad-tittle"
              >
                Silver Surfer
              </h5>
              <p
                className="card-text"
              >
                Norrin Radd
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  The Fantastic Four #48
                   
                </small>
              </p>
              <a
                href="/hero/marvel-silver"
                onClick={[Function]}
              >
                Ver más...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
